common/credentials/src/utils.rs

```diff
@@ -99,38 +112,64 @@ pub async fn obtain_aggregate_signature(
     let mut shares = Vec::with_capacity(validators.len());
 
     let mut client = validator_client::ApiClient::new(validators[0].clone());
-    let first =
-        obtain_partial_credential(params, public_attributes, private_attributes, &client).await?;
+    let validator_partial_vk = client.get_coconut_verification_key().await?;
+
+    let first = obtain_partial_credential(
+        params,
+        public_attributes,
+        private_attributes,
+        &client,
+        &validator_partial_vk.key,
+    )
+    .await?;
     shares.push(SignatureShare::new(first, 0));
 
     for (id, validator_url) in validators.iter().enumerate().skip(1) {
         client.change_validator_api(validator_url.clone());
-        let signature =
-            obtain_partial_credential(params, public_attributes, private_attributes, &client)
-                .await?;
+        let validator_partial_vk = client.get_coconut_verification_key().await?;
+        let signature = obtain_partial_credential(
+            params,
+            public_attributes,
+            private_attributes,
+            &client,
+            &validator_partial_vk.key,
+        )
+        .await?;
         let share = SignatureShare::new(signature, id as u64);
         shares.push(share)
     }
 
-    Ok(aggregate_signature_shares(&shares)?)
+    let mut attributes = Vec::with_capacity(private_attributes.len() + public_attributes.len());
+    attributes.extend_from_slice(private_attributes);
+    attributes.extend_from_slice(public_attributes);
+
+    Ok(aggregate_signature_shares(
+        params,
+        verification_key,
+        &attributes,
+        &shares,
+    )?)
 }
 
 // TODO: better type flow
 pub fn prepare_credential_for_spending(
     params: &Parameters,
     public_attributes: Vec<Vec<u8>>,
-    private_attributes: Vec<Vec<u8>>,
+    serial_number: Attribute,
+    binding_number: Attribute,
     signature: &Signature,
     verification_key: &VerificationKey,
 ) -> Result<Credential, Error> {
-    let private_attributes = private_attributes
-        .iter()
-        .map(hash_to_scalar)
-        .collect::<Vec<Attribute>>();
-    let theta = prove_credential(params, verification_key, signature, &private_attributes)?;
+    let theta = prove_bandwidth_credential(
+        params,
+        verification_key,
+        signature,
+        serial_number,
+        binding_number,
+    )?;
 
     Ok(Credential::new(
-        (public_attributes.len() + private_attributes.len()) as u32,
+        (public_attributes.len() + 2) as u32,
```

Comment:
It's 2 because each bandwidth voucher will have always 2 private attributes. I agree that this is not the most elegant way, I will replace it for now with a constant, but at some point, the code will need a bit of refactoring especially with the numeric numbers :)

Reply:
If this function is specific to bandwidth credential, maybe we should be passing `BandwidthVoucherAttributes` instead of all attributes as an argument instead? However, if it's something more generic then yeah, it doesn't make sense.